### PR TITLE
24 add cross attention labels text

### LINKED ISSUE #24: Add Cross Attention Labels / Text
Implement cross attention between labels and text embedding.

Examples:
[Here](https://ieeexplore.ieee.org/stamp/stamp.jsp?tp=&arnumber=10409522), they call it Label Attention (queries are labels and key/value are text embedding, _the labels attend to tokens_ to get _the label-attend text repr._ )

[Here](https://mqianliu.github.io/files/CNLE_Neurocomputing22.pdf), they compute also _the text-attended label representations_ (query = text embedding, key/value = labels)...

This would enable another explainability pipeline, that would be independent from Captum, above all internal to the model structure itself.

### PR DESCRIPTION
**Pull request by meilame-tayebjee**
This pull request introduces label attention as an optional feature in the text classification pipeline, allowing the model to generate label-specific sentence embeddings using a cross-attention mechanism. The changes include new configuration classes, updates to the `TextEmbedder` and model logic, and new tests to ensure label attention works as intended.

**Label Attention Mechanism:**

* Added `LabelAttentionConfig` and `LabelAttentionClassifier` to enable label-specific sentence embeddings using cross-attention, where labels act as queries over token embeddings. [[1]](diffhunk://#diff-c943288e8a3126107bb470078b3627a8b2dfea92ec06ca0d678140aad8bdc924L6-R25) [[2]](diffhunk://#diff-c943288e8a3126107bb470078b3627a8b2dfea92ec06ca0d678140aad8bdc924R262-R340)
* Updated `TextEmbedderConfig` and `TextEmbedder` to support label attention, including a new output structure and logic to handle label attention matrices. [[1]](diffhunk://#diff-c943288e8a3126107bb470078b3627a8b2dfea92ec06ca0d678140aad8bdc924L6-R25) [[2]](diffhunk://#diff-c943288e8a3126107bb470078b3627a8b2dfea92ec06ca0d678140aad8bdc924L29-R40) [[3]](diffhunk://#diff-c943288e8a3126107bb470078b3627a8b2dfea92ec06ca0d678140aad8bdc924R53) [[4]](diffhunk://#diff-c943288e8a3126107bb470078b3627a8b2dfea92ec06ca0d678140aad8bdc924L108-R135) [[5]](diffhunk://#diff-c943288e8a3126107bb470078b3627a8b2dfea92ec06ca0d678140aad8bdc924L141-R182) [[6]](diffhunk://#diff-c943288e8a3126107bb470078b3627a8b2dfea92ec06ca0d678140aad8bdc924L166-R200) [[7]](diffhunk://#diff-c943288e8a3126107bb470078b3627a8b2dfea92ec06ca0d678140aad8bdc924R218-R240)

**Model and Pipeline Integration:**

* Modified the main model (`model.py`) to validate and propagate label attention configuration, including enforcing that the classification head outputs a single value when label attention is enabled and updating the number of classes accordingly.
* Adjusted the forward method and related pipeline logic to support returning label attention matrices and to handle the new embedding shapes.

**Testing Enhancements:**

* Updated and extended the test pipeline to support label attention, including a new test function `test_label_attention_enabled` and corresponding updates to the helper functions. [[1]](diffhunk://#diff-ef5f0938c9abd7d9ab658f28e363b093e59f677949829145bc7d6036cec3897fR12) [[2]](diffhunk://#diff-ef5f0938c9abd7d9ab658f28e363b093e59f677949829145bc7d6036cec3897fL54-R62) [[3]](diffhunk://#diff-ef5f0938c9abd7d9ab658f28e363b093e59f677949829145bc7d6036cec3897fR94-R102) [[4]](diffhunk://#diff-ef5f0938c9abd7d9ab658f28e363b093e59f677949829145bc7d6036cec3897fL101-R118) [[5]](diffhunk://#diff-ef5f0938c9abd7d9ab658f28e363b093e59f677949829145bc7d6036cec3897fR156-R164) [[6]](diffhunk://#diff-ef5f0938c9abd7d9ab658f28e363b093e59f677949829145bc7d6036cec3897fL166-R204) [[7]](diffhunk://#diff-ef5f0938c9abd7d9ab658f28e363b093e59f677949829145bc7d6036cec3897fR274-R296)

**Miscellaneous:**

* Minor code improvements, such as import adjustments and docstring updates to reflect the new model name and features. [[1]](diffhunk://#diff-565a2b7fb641c72439fa6f5a9cc66073cc7dd292085fb1735887a5dc1cf21cd6R11) [[2]](diffhunk://#diff-aaeab4afeaf4bf324ccf781f3530d5bd78bf8a47e6bb902887daf62de0124470L1-R4) [[3]](diffhunk://#diff-aaeab4afeaf4bf324ccf781f3530d5bd78bf8a47e6bb902887daf62de0124470R20) [[4]](diffhunk://#diff-aaeab4afeaf4bf324ccf781f3530d5bd78bf8a47e6bb902887daf62de0124470L70-L71) [[5]](diffhunk://#diff-e97a297a4a58822e9c33f52eebd7bf8ad622bef3e0d108bc605155b96995d1e8R105)

### CODE REVIEW

File: torchTextClassifiers/model/components/text_embedder.py [hunk 1]
```diff
@@ -221,3 +259,82 @@ def _precompute_rotary_embeddings(self, seq_len, head_dim, base=10000, device=No
...
+            self.n_head != self.n_kv_head
+        )  # Group Query Attention (GQA): duplicate key/value heads to match query heads if desired
+        self.head_dim = self.embedding_dim // self.n_head
+
+        self.label_embeds = nn.Embedding(self.num_classes, self.embedding_dim)
+
+        self.c_q = nn.Linear(self.embedding_dim, self.n_head * self.head_dim, bias=False)
+        self.c_k = nn.Linear(self.embedding_dim, self.n_kv_head * self.head_dim, bias=False)
+        self.c_v = nn.Linear(self.embedding_dim, self.n_kv_head * self.head_dim, bias=False)
+        self.c_proj = nn.Linear(self.embedding_dim, self.embedding_dim, bias=False)
```
[the remark this answers is not in the document]
Comment (meilame-tayebjee) [reply torchTextClassifiers/model/components/text_embedder.py:323]: @copilot open a new pull request to apply changes based on [this feedback](https://github.com/InseeFrLab/torchTextClassifiers/pull/60#discussion_r2728535183)

File: torchTextClassifiers/model/components/text_embedder.py [hunk 2]
```diff
@@ -221,3 +259,82 @@ def _precompute_rotary_embeddings(self, seq_len, head_dim, base=10000, device=No
...
+        if isinstance(compute_attention_matrix, torch.Tensor):
+            compute_attention_matrix = compute_attention_matrix[0].item()
+        compute_attention_matrix = bool(compute_attention_matrix)
+
+        # 1. Create label indices [0, 1, ..., C-1] for the whole batch
+        label_indices = torch.arange(self.num_classes).expand(B, -1)
+
+        all_label_embeddings = self.label_embeds(
+            label_indices
+        )  # Shape: [batch, num_classes, d_model]
```
[the remark this answers is not in the document]
Comment (meilame-tayebjee) [reply torchTextClassifiers/model/components/text_embedder.py:355]: @copilot open a new pull request to apply changes based on [this feedback](https://github.com/InseeFrLab/torchTextClassifiers/pull/60#discussion_r2728535163)

File: torchTextClassifiers/model/components/text_embedder.py [hunk 3]
```diff
@@ -181,25 +215,29 @@ def _get_sentence_embedding(
...
-
-    def __call__(self, *args, **kwargs):
-        out = super().__call__(*args, **kwargs)
-        if out.dim() != 2:
-            raise ValueError(
-                f"Output of {self.__class__.__name__}.forward must be 2D "
-                f"(got shape {tuple(out.shape)})"
+        if self.enable_label_attention:
+            label_attention_result = self.label_attention_module(
+                token_embeddings, compute_attention_matrix=return_label_attention_matrix
```
[the remark this answers is not in the document]
Comment (meilame-tayebjee) [reply torchTextClassifiers/model/components/text_embedder.py]: @copilot thanks for the catch. You have changed args but have you really masked the attention ?

File: torchTextClassifiers/model/components/text_embedder.py [hunk 4]
```diff
@@ -221,3 +279,118 @@ def _precompute_rotary_embeddings(self, seq_len, head_dim, base=10000, device=No
...
+        attention_matrix = None
+        if compute_attention_matrix:
+            # Compute attention scores
+            # size (B, n_head, n_labels, seq_len)
+            attention_scores = torch.matmul(q, k.transpose(-2, -1)) / (self.head_dim**0.5)
+
+            # Apply mask to attention scores before softmax
+            if attention_mask is not None:
+                # attn_mask is already in the right shape: (B, 1, 1, T)
+                # We need to apply it to scores of shape (B, n_head, n_labels, T)
```
[the remark this answers is not in the document]
Comment (meilame-tayebjee) [reply torchTextClassifiers/model/components/text_embedder.py:395]: would rather remove GQA for label attention to simplify for now @copilot

File: torchTextClassifiers/model/model.py [hunk 5]
```diff
@@ -108,6 +118,7 @@ def forward(
         input_ids: Annotated[torch.Tensor, "batch seq_len"],
         attention_mask: Annotated[torch.Tensor, "batch seq_len"],
         categorical_vars: Annotated[torch.Tensor, "batch num_cats"],
+        return_label_attention_matrix: bool = False,
         **kwargs,
     ) -> torch.Tensor:
```
[the remark this answers is not in the document]
Comment (meilame-tayebjee) [reply torchTextClassifiers/model/model.py]: @copilot open a new pull request to apply changes based on [this feedback](https://github.com/InseeFrLab/torchTextClassifiers/pull/60#discussion_r2731219927)

File: torchTextClassifiers/model/model.py [hunk 6]
```diff
@@ -126,26 +137,46 @@ def forward(
...
+
                 x_combined = x_text + x_cat
         else:
             x_combined = x_text
 
-        logits = self.classification_head(x_combined)
+        logits = self.classification_head(norm(x_combined)).squeeze(-1)
+
+        if return_label_attention_matrix:
+            return {"logits": logits, "label_attention_matrix": label_attention_matrix}
```
[the remark this answers is not in the document]
Comment (meilame-tayebjee) [reply torchTextClassifiers/model/model.py:190]: @copilot open a new pull request to apply changes based on [this feedback](https://github.com/InseeFrLab/torchTextClassifiers/pull/60#discussion_r2731219946)

File: torchTextClassifiers/model/components/text_embedder.py [hunk 7]
```diff
@@ -163,7 +215,7 @@ def _get_sentence_embedding(
         # mask pad-tokens
 
         if self.attention_config is not None:
-            if self.attention_config.aggregation_method is not None:
+            if self.attention_config.aggregation_method is not None:  # default is "mean"
                 if self.attention_config.aggregation_method == "first":
                     return token_embeddings[:, 0, :]
```
[the remark this answers is not in the document]
Comment (meilame-tayebjee) [reply torchTextClassifiers/model/components/text_embedder.py]: @copilot open a new pull request to apply changes based on [this feedback](https://github.com/InseeFrLab/torchTextClassifiers/pull/60#discussion_r2731219958)

File: torchTextClassifiers/model/components/text_embedder.py [hunk 8]
```diff
@@ -138,14 +180,24 @@ def forward(self, input_ids: torch.Tensor, attention_mask: torch.Tensor) -> torc
...
+            "label_attention_matrix": label_attention_matrix,
+        }
 
     def _get_sentence_embedding(
-        self, token_embeddings: torch.Tensor, attention_mask: torch.Tensor
+        self,
+        token_embeddings: torch.Tensor,
+        attention_mask: torch.Tensor,
+        return_label_attention_matrix: bool = False,
     ) -> torch.Tensor:
```
[the remark this answers is not in the document]
Comment (meilame-tayebjee) [reply torchTextClassifiers/model/components/text_embedder.py]: @copilot open a new pull request to apply changes based on [this feedback](https://github.com/InseeFrLab/torchTextClassifiers/pull/60#discussion_r2731219976)

File: tests/test_pipeline.py [hunk 9]
```diff
@@ -163,13 +189,19 @@ def run_full_pipeline(tokenizer, sample_text_data, categorical_data, labels, mod
 
     # Predict with explanations
     top_k = 5
-    predictions = ttc.predict(X, top_k=top_k, explain=True)
+
+    predictions = ttc.predict(
+        X,
+        top_k=top_k,
+        explain_with_label_attention=label_attention_enabled,
+        explain_with_captum=True,
```
[the remark this answers is not in the document]
Comment (meilame-tayebjee) [reply tests/test_pipeline.py:195]: @copilot open a new pull request to apply changes based on [this feedback](https://github.com/InseeFrLab/torchTextClassifiers/pull/60#discussion_r2731220040)
